sidecar/cook/sidecar/progress.py

```diff
@@ -0,0 +1,127 @@
+            except Exception:
+                logging.exception(f'Error raised while posting progress update to {current_url}')
+            current_url = default_url
+            logging.info(f'Failed to post progress update. Reset progress update callback url: {current_url}')
+            return False
+
+        max_message_length = config.max_message_length
+        sample_interval_ms = config.progress_sample_interval_ms
+        sequence_counter = cst.ProgressSequenceCounter()
+        progress_updater = cst.ProgressUpdater(max_message_length, sample_interval_ms, send_progress_message)
+
+        def launch_progress_tracker(progress_location, location_tag):
+            logging.info(f'Location {progress_location} tagged as [tag={location_tag}]')
+            progress_tracker = cst.ProgressTracker(config, sequence_counter, progress_updater, progress_location, location_tag)
+            progress_tracker.start()
+            return progress_tracker
+
+        progress_locations = {config.progress_output_name: 'progress',
+                              config.stderr_file(): 'stderr',
+                              config.stdout_file(): 'stdout'}
+        logging.info(f'Progress will be tracked from {len(progress_locations)} locations')
+        progress_trackers = [launch_progress_tracker(file, name) for file, name in progress_locations.items()]
+
+        def handle_interrupt(interrupt_code, _):
+            logging.info(f'Progress Reporter interrupted with code {interrupt_code}')
+            # force send the latest progress state if available, and stop the tracker
+            for progress_tracker in progress_trackers:
+                progress_tracker.force_send_progress_update()
+                progress_tracker.stop()
+
+        signal.signal(signal.SIGINT, handle_interrupt)
+        signal.signal(signal.SIGTERM, handle_interrupt)
+
+        def dump_traceback(signal, frame):
+            faulthandler.dump_traceback()
+
+        signal.signal(signal.SIGUSR1, dump_traceback)
+
+        return progress_trackers
+
+    except Exception:
+        logging.exception('Error starting Progress Reporter')
+        return None
+
+
+def await_progress_trackers(progress_trackers):
+        if progress_trackers is None:
+            sys.exit('Failed to start progress trackers')
+        # wait for all background threads to exit
+        # (but this process will probably be killed first instead)
+        for progress_tracker in progress_trackers:
+            progress_tracker.wait()
+
+
+def main():
+    log_level = os.environ.get('EXECUTOR_LOG_LEVEL', 'INFO')
+    logging.basicConfig(level = log_level,
+                        stream = sys.stderr,
+                        format='%(asctime)s %(levelname)s %(message)s')
+    if len(sys.argv) == 2 and sys.argv[1] == "--version":
+        print(__version__)
+    else:
+        logging.info(f'Starting cook.sidecar {__version__} progress reporter')
```

Comment:
Please move this log to `start_progress_trackers`

Reply:
:heavy_check_mark: